File: src/jsonrpc/eth/getBlockByNumber/eth_getBlockByNumber.zig
const std = @import("std");
const types = @import("../../types.zig");

/// Returns information about a block by number.
///
/// Example:
/// block: "0x68b3"
/// Hydrated transactions: false
/// Result: ...
///
/// Implements the `eth_getBlockByNumber` JSON-RPC method.
pub const EthGetBlockByNumber = @This();

/// The JSON-RPC method name
pub const method = "eth_getBlockByNumber";

/// Parameters for `eth_getBlockByNumber`
pub const Params = struct {
    /// Block number or tag
    block: types.Quantity,
    /// hydrated
    hydrated transactions: types.Quantity,

    pub fn jsonStringify(self: Params, jws: *std.json.Stringify) !void {
        try jws.beginArray();
        try jws.write(self.block);
        try jws.write(self.hydrated transactions);
        try jws.endArray();
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Params {
        if (source != .array) return error.UnexpectedToken;
        if (source.array.items.len != 2) return error.InvalidParamCount;

        return Params{
            .block = try std.json.innerParseFromValue(types.Quantity, allocator, source.array.items[0], options),
            .hydrated transactions = try std.json.innerParseFromValue(types.Quantity, allocator, source.array.items[1], options),
        };
    }
};

/// Result for `eth_getBlockByNumber`
pub const Result = struct {
    value: types.Quantity,

    pub fn jsonStringify(self: Result, jws: *std.json.Stringify) !void {
        try jws.write(self.value);
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Result {
        return Result{
            .value = try std.json.innerParseFromValue(types.Quantity, allocator, source, options),
        };
    }
};

File: src/jsonrpc/eth/getBlockByNumber/eth_getBlockByNumber.js
/**
 * @fileoverview eth_getBlockByNumber JSON-RPC method
 */

/**
 * @typedef {import('../../../primitives/Address/AddressType.js').AddressType} Address
 * @typedef {import('../../index.js').Hash} Hash
 * @typedef {import('../../index.js').Quantity} Quantity
 * @typedef {import('../../index.js').BlockTag} BlockTag
 * @typedef {import('../../index.js').BlockSpec} BlockSpec
 * @typedef {import('../../../provider/types.js').RequestArguments} RequestArguments
 */

/**
 * Returns information about a block by number.
 *
 * @example
 * block: "0x68b3"
 * Hydrated transactions: false
 * Result: ...
 *
 * Implements the `eth_getBlockByNumber` JSON-RPC method.
 */

/** The JSON-RPC method name */
export const method = "eth_getBlockByNumber";
/**
 * Result for `eth_getBlockByNumber`
 *
 * @typedef {Quantity} Result
 */

/**
 * Creates a eth_getBlockByNumber JSON-RPC request
 *
 * @param {BlockSpec} block - Block number, tag, or block hash
 * @param {boolean} [fullTransactions=false] - If true, returns full transaction objects
 * @returns {RequestArguments}
 */
export function GetBlockByNumberRequest(block, fullTransactions = false) {
	return { method, params: [block, fullTransactions] };
}
